pkg/virt-controller/watch/export/export.go

```diff
@@ -384,19 +393,147 @@ func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMach
+		}
+	}
+
+	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
+
+}
+
+func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) (time.Duration, error) {
+	var retry time.Duration
+	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
+
+	vmSnapshot, exists, err := ctrl.getVmSnapshot(vmExport.Namespace, vmExport.Spec.Source.Name)
+	restoreableSnapshots := 0
+	if err != nil {
+		return 0, err
+	} else if exists {
+		pvcs, restoreableSnapshots, err = ctrl.handlePVCsForVirtualMachineSnapshot(vmExport, vmSnapshot)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	pod, exists, err := ctrl.getExporterPod(vmExport)
+	if err != nil {
+		return 0, err
+	} else if !exists {
+		if len(pvcs) == restoreableSnapshots && restoreableSnapshots > 0 {
+			pod, err = ctrl.createExporterPod(vmExport, pvcs)
+			if err != nil {
 				return 0, err
-			} else {
+			} else if pod == nil {
 				return retry, nil
 			}
+
+			if err := ctrl.getOrCreateCertSecret(vmExport, pod); err != nil {
+				return 0, err
+			}
+		}
+	} else {
+		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+			return retry, err
 		}
 	}
 
-	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
+	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, false)
+}
+
+func (ctrl *VMExportController) handlePodSucceededOrFailed(vmExport *exportv1.VirtualMachineExport, pod *corev1.Pod) error {
+	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+		// The server died or completed, delete the pod.
+		ctrl.Recorder.Eventf(vmExport, corev1.EventTypeWarning, exporterPodFailedOrCompletedEvent, "Exporter pod %s/%s succeeded or failed", pod.Namespace, pod.Name)
+		if err := ctrl.Client.CoreV1().Pods(vmExport.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{}); err != nil {
+			return err
+		} else {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (ctrl *VMExportController) handlePVCsForVirtualMachineSnapshot(vmExport *exportv1.VirtualMachineExport, vmSnapshot *snapshotv1.VirtualMachineSnapshot) ([]*corev1.PersistentVolumeClaim, int, error) {
+	var content *snapshotv1.VirtualMachineSnapshotContent
+	var err error
+	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
+	exists := false
+	totalVolumes := 0
+
+	if vmSnapshot.Status.VirtualMachineSnapshotContentName != nil && *vmSnapshot.Status.VirtualMachineSnapshotContentName != "" {
+		content, exists, err = ctrl.getVmSnapshotContent(vmSnapshot.Namespace, *vmSnapshot.Status.VirtualMachineSnapshotContentName)
+		if err != nil {
+			return pvcs, totalVolumes, err
+		} else if exists {
+			sourceVm := content.Spec.Source.VirtualMachine
+			totalVolumes = len(content.Status.VolumeSnapshotStatus)
+			for _, volumeBackup := range content.Spec.VolumeBackups {
+				if pvc, err := ctrl.getOrCreatePVCFromSnapshot(vmExport, &volumeBackup, sourceVm); err != nil {
+					return pvcs, totalVolumes, err
```

Comment:
does it need to fail if only one of the PVCs returns an error? We are ignoring the following ones

Reply:
Yes, we only want to start the export if all the pvcs are available since we are mounting them all in the exporter pod.

Reply:
ok, but you are returning some partial results for `pvcs` and `totalVolumes` if there is an error. Maybe you could return 0and an empty slice

Reply:
Yeah it's kind of weird to return partial results if there is an error

Reply:
Well the caller will ignore the value on the error, but I can return an empty slice instead. Doesn't matter that much.